src/test/kotlin/cacophony/semantic/TypeCheckerTest.kt

```diff
@@ -0,0 +1,952 @@
+@file:Suppress("ktlint")
+
+package cacophony.semantic
+
+import cacophony.semantic.syntaxtree.*
+import cacophony.utils.Diagnostics
+import cacophony.utils.Location
+import org.junit.jupiter.api.Assertions.*
+import org.junit.jupiter.api.Test
+
+class TypeCheckerTest {
+    private fun assertTypeEquals(
+        type1: TypeExpr?,
+        type2: TypeExpr?,
+    ) = assertEquals(type1.toString(), type2.toString())
+
+    @Test
+    fun `equality check`() {
+        val type1 = BuiltinType.BooleanType
+        val type2 = BuiltinType.BooleanType
+        assertEquals(type1, type2)
+    }
+
+    @Test
+    fun `non-equality check`() {
+        val type1 = BuiltinType.BooleanType
+        val type2 = BuiltinType.IntegerType
+        assertNotEquals(type1, type2)
+    }
+
+    @Test
+    fun `function type string representation`() {
+        val functionType = FunctionType(listOf(BuiltinType.BooleanType, BuiltinType.IntegerType), BuiltinType.UnitType)
+        assertEquals("FunctionType(args=[Boolean, Int], result=Unit)", functionType.toString())
+        assertEquals("[Boolean, Int] -> Unit", functionType.name)
+    }
+
+    private val lc = Pair(Location(0), Location(0))
+
+    class DiagnosticsMock : Diagnostics {
+        var msg: String? = null
+
+        override fun report(
+            message: String,
+            location: Location,
+        ) {
+            TODO()
+        }
+
+        override fun report(
+            message: String,
+            range: Pair<Location, Location>,
+        ) {
+            msg = message
+        }
+    }
+
+    private fun getDiagnostic(): DiagnosticsMock = DiagnosticsMock()
+
+    private val testUnit = Type.Basic(lc, "Unit")
+    private val testInt = Type.Basic(lc, "Int")
+    private val testBoolean = Type.Basic(lc, "Boolean")
+    private val intLiteral = Literal.IntLiteral(lc, 7)
+    private val booleanLiteral = Literal.BoolLiteral(lc, true)
+
+    @Test
+    fun `ok - empty block`() {
+        val ast = Block(lc, emptyList())
+        val diagnostics = getDiagnostic()
+        val result = checkTypes(ast, diagnostics, emptyMap())
+        assertTypeEquals(BuiltinType.UnitType, result[ast])
+        assertNull(diagnostics.msg)
```

Comment:
you can do

```
@MockK
lateinit var diagnostics: Diagnostics

@BeforeEach
fun setUpMocks() {
    MockKAnnotations.init(this, relaxUnitFun = true)
    every { diagnostics.report(any(), any<Location>()) } just runs
}
```

and then in tests

```
verify(exactly = 0) { diagnostics.report(any(), any<Location>()) }
```

like in https://github.com/mhorod/cacophony/blob/53493d26df99c0c86e1fb7beedbe9c2536f554aa/src/test/kotlin/cacophony/lexer/RegularLanguageLexerTest.kt#L98

Reply:
I can, but should I?

Reply:
up to you, I prefer it this way but its a matter of taste